---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# typGumbo

<!-- badges: start -->
<!-- badges: end -->

The goal of typGumbo is to ...

## Installation

typGumbo is not on CRAN so it can NOT be loaded via library("typGumbo")

The development version of typGumbo can be obtained from [GitHub](https://github.com/) with:

``` {r}
# install.packages("devtools")
devtools::install_github("genejockey33000/typGumbo")
```

## Example
This is a basic example which shows you how use a few of the functions:
```{r example}
library(typGumbo)
## Generate a sleuth Object from kallisto output
so <- run.sleuth(in.dir = "results", meta = "meta", level = "gene")

## Generate a "typ" class R object for use with other Gumbo functions
#from a sleuth object (meta data and level are pulled automatically)
obj <- make.typ(so, regress = c("RNAseqLibBatch", "diffRound"))

#from a sleuth object and regressing for sequencing round and differentiation batch
obj <- make.typ(so)

#from a matrix (measurements in rows)
obj <- make.typ(x = {inputmatrix}, level = "gene", meta = meta, regress = c("RNAseqLibBatch", "diffRound"))
```


You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this. You could also use GitHub Actions to re-render `README.Rmd` every time you push. An example workflow can be found here: <https://github.com/r-lib/actions/tree/master/examples>.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub and CRAN.
